File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.ts
function maxProfit(prices: number[]): number {
    const n = prices.length
    const left = Array(n).fill(0)
    let min = prices[0]
    for (let i=1; i < n; i++) {
        left[i] = Math.max(left[i-1], prices[i] - min)
        min = Math.min(min, prices[i])
    }
    const right = Array(n).fill(0)
    let max = prices[n-1]
    for (let i=n-2; i >= 0; i--) {
        right[i] = Math.max(right[i+1], max - prices[i])
        max = Math.max(max, prices[i])
    }
    let res = -Infinity
    for (let i=0; i < n; i++) {
        res = Math.max(res, left[i] + right[i])
    }
    return res
};

File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        def k_transactions(k: int, prices: List[int]) -> int:
        
            dp = [[0] * len(prices) for _ in range(k+1)]
            
            for i in range(1, k+1):
                temp = -inf
                for j in range(1, len(prices)):
                    temp = max(temp, dp[i-1][j-1] - prices[j-1])
                    dp[i][j] = max(dp[i][j-1], temp + prices[j])
                    
            return dp[k][len(prices)-1]
        
        return k_transactions(2, prices)